macrotranslator.py

```diff
@@ -0,0 +1,87 @@
+from macros import MacroMap, Macro, Invocation
+from translationconfig import TranslationConfig
+import logging
+import re
+
+logger = logging.getLogger(__name__)
+
+
+class MacroTranslator:
+    def __init__(self, translation_config: TranslationConfig) -> None:
+        self.translation_config = translation_config
+
+    def generate_macro_translations(self,
+                                    mm: MacroMap) -> dict[Macro, str | None]:
+        translationMap: dict[Macro, str | None] = {}
+
+        for macro, invocations in mm.items():
+            translationMap[macro] = self.get_macro_translation(macro, invocations)
+
+        return translationMap
+
+    def get_macro_translation(self, macro: Macro, invocations: set[Invocation]) -> str | None:
+
+        if self.should_skip_macro(macro, invocations):
+            return None
+
+        if macro.IsFunctionLike:
+            return self.translate_function_like_macro(macro, invocations)
+        elif macro.IsObjectLike:
+            return self.translate_object_like_macro(macro, invocations)
+
+    def should_skip_macro(self, macro: Macro, invocations: set[Invocation]) -> bool:
+        # For now, skip translating anything with a function pointer type
+        # As Maki does not output the correct C syntax for these
+
+        # If any part of the type signature has a function pointer
+        invocation_has_function_type = \
+            any(i.IsExpansionTypeFunctionType or i.IsAnyArgumentTypeFunctionType for i in invocations)
+
+        # TODO(Joey): Implement a way to translate these
+        if invocation_has_function_type:
+            logger.debug(f"Skipping {macro.Name} as it has a function pointer type")
+            return True
+
+        # If body contains a DeclRefExpr and is in a header file, skip
+        # TODO(Joey/Brent): Find better way on Maki side to handle this
+        invocation = next(iter(invocations))
+
+        invocation_has_decl_ref_expr = invocation.DoesBodyContainDeclRefExpr
+        if invocation_has_decl_ref_expr and invocation.DefinitionLocationFilename.endswith(".h"):
+            logger.debug(f"Skipping {macro.Name} as it contains a DeclRefExpr")
+            return True
+
+        return False
+
+    def translate_function_like_macro(self, macro: Macro, invocations: set[Invocation]) -> str:
+        # Make sure we don't return for void functions,
+        # but do return for void * and anything else
+        pattern = r"void(?!\s*\*)"
```

Comment:
I think you can use Maki's `IsExpansionTypeVoid` property here to perform this check more robustly and without needing to use regex.

Reply:
Good catch, updated.